background/objectoriented.tex

```diff
@@ -1,210 +1,333 @@
-This implies for instance that the control registers for a devices such as an Analog-To-Digital is located in the address space of the CPU and can be read from and written to with the same instruction and language constructs as the rest of the memory.
-\autoref{fig:back:memorymapped} shows the control registers for the device which are placed continuously in memory.
-
+In its essense Object Oriented programming focues on organizing a computer program by looking at the data the program operates on, grouping related data into objects and defining methods that operate on the data contained within the objects.
+The paradigms essensial concept be applied to embedded C programming even though the language does not define language features to aid the design.
+This section looks at how Object Oriented programming is used in conjunction with memory mapped I/O in embedded programming.
+We look at the memory layout of a memory mapped Analoge-To-Digital and see how this can be represented as an object.
+
+\subsection{Memory Mapped I/O}
+Memory Mapped I/0, MMIO is a method for interfacing with peripheral devices in a computer system.
+The method entailes connecting the control registers of hardware devices to the same address bus as RAM.
+This results in a programming model where the programmer can use common memory operations to control the devices.
+
+Lets consider the Analoge-To-Digital, ADC, converter on the Gecko.
+The ADC converts an analoge signal to a digital representation.
+The ADC on the Gecko is memory mapped to the location \textbf{0x40002000}.
+This means that writing to a pointer pointing to this address will write to the control registers in the ADC device.
+A subsection of the ADC control register layout is given in \autoref{tab:back:memorymapped}.
 \begin{figure}[H]
   \centering
   \begin{tabular}{l|l|l|}
-    &&\\
-    \hline
     Location&Offset&Name\\
     \hline
+    &...&...\\
+    \hline
+    \hline
     0x40002000&0x0&CTRL\\
     \hline
     &0x4&CMD\\
     \hline
-    &0x8&STATUS\\
+    &...&...\\
+    \hline
+    &0xC&SINGLECTRL\\
     \hline
-    &...&... \\
+    &...&...\\
+    \hline
+    &0x24&SINGLEDATA\\
+    \hline
+    &...&...\\
     \hline
-    &0x3C&BIASPROG\\
     \hline
     &&\\
-
   \end{tabular}
-  \caption{ADC0 Memory map for EFM32GG}
+  \caption{Subsection of ADC0 Memory map for EFM32GG}
   \label{tab:back:memorymapped}
 \end{figure}
 
-\autoref{fig:back:objectmemory} shows the memory layout for the objects defined in \autoref{sec:back:oo}.
-We see that the difference between the Java managed language and the C and Rust system langauges is the tag field preciding the data fields for the object.
+\autoref{tab:back:memorymapped} shows the control register needed to perform a single ADC convertion.
+The ADC control register contains more register than in \autoref{tab:back:memorymapped}, butall registers in the ADC register file is aligned on a 4 byte boundry.
```

Comment:
Trenger kanskje ikke referere til figure 3 ganger på 3 setninger. 
f.eks. no a-la

A subsection of the ADC control register layout is given in `figure`.
It shows the control register needed to perform a single ADC conversion, note that it only shows the registers needed to perform a single ADC conversion.